swig/ChunkedArray.hpp

```diff
@@ -0,0 +1,240 @@
+#ifndef __CHUNKED_ARRAY_H__
+#define __CHUNKED_ARRAY_H__
+
+#include <new>
+#include <vector>
+#include <algorithm>
+
+#include <stdint.h>
+#include <assert.h>
+
+/**
+ * Container that manages a dynamic array of fixed-length chunks.
+ *
+ * The class also takes care of allocation & release of the underlying
+ * memory. It can be used with either a high or low-level API.
+ *
+ * The high-level API allocates chunks as needed, manages addresses automatically and keeps
+ * track of number of inserted elements, but is not thread-safe (ok as usually input is a streaming iterator).
+ * For parallel input sources the low-level API must be used.
+ *
+ * Note: When using this for `LGBM_DatasetCreateFromMats` use a
+ *       chunk_size multiple of #num_cols for your dataset, so each chunk
+ *       contains "complete" instances.
+ *
+ * === High-level insert API intro ===
+ *
+ * The easiest way to use is:
+ *  0. ChunkedArray(chunk_size)  # Choose appropriate size
+ *  1. add(value)                # as many times as you want (will generate chunks as needed)
+ *  2. data() or void_data()     # retrieves a T** or void** pointer (useful for `LGBM_DatasetCreateFromMats`).
+ *
+ * Useful query methods (all O(1)):
+ *  - get_add_count()   # total count of added elements
+ *  - get_chunks_count()  # how many chunks are currently allocated.
+ *  - get_current_chunk_added_count()  # for the last add() chunk, how many items there are.
+ *  - get_chunk_size()    # Get constant chunk_size from constructor call.
+ *
+ *  With those you can generate int32_t sizes[]. Last chunk can be smaller than chunk_size, so, for any i:
+ *    - sizes[i<last] = get_chunk_size()
+ *    - sizes[i==last] = get_add_count()
+ *
+ *
+ * === Low-level insert API intro ===
+ *
+ * For advanced usage - useful for inserting in parallel - one can also:
+ *  1. call new_chunk() at any time for as many chunks as needed.  (thread-UNsafe)
+ *  2. call setitem(chunk, idx, value) to insert each value.       (thread-safe)
+ *
+ */
+template <class T>
+class ChunkedArray
+{
+  public:
+    ChunkedArray(size_t chunk_size)
+      : _chunk_size(chunk_size), _current_chunks_idx(0), _current_chunk_idx(0)
```

Comment:
these names look too similar:
_current_chunks_idx
_current_chunk_idx
what do you think about changing latter to:
_current_in_chunk_idx or _current_within_chunk_idx
just worried it's so similar that it's easy to make mistake

Reply:
Yeah I find them confusing too :p Continuing the logic from your suggestion:

What about `_last_chunk_idx` and `_last_idx_in_last_chunk`?